lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,210 @@
+require 'uri'
+
+module ProxyAPI
+  class ExternalIpam < ProxyAPI::Resource
+    def initialize(args)
+      @url = args[:url] + "/ipam"
+      super args
+    end
+
+    # Queries External IPAM and retrieves the next available IP address for the given subnet.
+    # The IP returned is NOT reserved in External IPAM database. It is however written to an in-memory,
+    # thread safe cache, on the proxy side, with the subnet CIDR/mac address as the key(see
+    # IP cache structure below). This will prevent the same IP being suggested for different mac addresses,
+    # and will handle race condition scenarios where multiple hosts are being provisioned simultaneously.
+    #
+    # The IP address is only actually reserved in the External IPAM database upon successful host
+    # and/or interface creation.
+    #
+    # In-memory IP cache structure(IPv4 and IPv6):
+    # ===============================
+    #   "100.55.55.0/24":{
+    #      "00:0a:95:9d:68:10": {"ip": "100.55.55.1", "timestamp": "2019-09-17 12:03:43 -D400"}
+    #   },
+    #   "123.11.33.0/24":{
+    #      "00:0a:95:9d:68:33": {"ip": "123.11.33.1", "timestamp": "2019-09-17 12:04:43 -0400"},
+    #      "00:0a:95:9d:68:34": {"ip": "123.11.33.2", "timestamp": "2019-09-17 12:05:48 -0400"},
+    #      "00:0a:95:9d:68:35": {"ip": "123.11.33.3", "timestamp:: "2019-09-17 12:06:50 -0400"}
+    #   },
+    #   "2001:db8:abcd:12::/124":{
+    #      "00:0a:95:9d:68:10": {"ip": "2001:db8:abcd:12::1", "timestamp": "2019-09-17 12:03:43 -D400"},
+    #      "00:0a:95:9d:68:11": {"ip": "2001:db8:abcd:12::2", "timestamp": "2019-09-17 12:13:22 -D400"}
+    #   }
+    # }
+    #
+    # Params: 1. subnet:  The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24", IPv6 - "2001:db8:abcd:12::/124")
+    #         2. mac:     The mac address of the interface obtaining the IP address for
+    #
+    # Returns: A hash with next available IP in the "data" field, or a hash with "error" on failure
+    #
+    # Examples:
+    #   Responses if success:
+    #     IPv4: {"code": 200, "success": true, "data": "100.55.55.3", "time": 0.012}
+    #     IPv6: {"code": 200, "success": true, "data": "2001:db8:abcd:12::1", "time": 0.012}
+    #   Response if missing required params:
+    #     {"error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","A 'mac' address must be provided(e.g. 00:0a:95:9d:68:10)"]}
+    #   Response if subnet error:
+    #     {"error": "The specified subnet does not exist in External IPAM."}
+    #   Response if there are no free addresses:
+    #     {"error": "No free addresses found"}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def next_ip(subnet, mac)
+      raise "subnet cannot be nil" if subnet.nil?
+      parse get("/next_ip?cidr=#{subnet}&mac=#{mac}")
```

Comment:
I think this also works and gives you escaping of arguments (but untested):

```suggestion
      parse get("/next_ip, subnet: subnet, mac: mac)
```

Reply:
Tried this but unfortunately didn't work